crypto/cms/cms_env.c

```diff
@@ -53,19 +54,60 @@ static CMS_EnvelopedData *cms_enveloped_data_init(CMS_ContentInfo *cms)
 int cms_env_asn1_ctrl(CMS_RecipientInfo *ri, int cmd)
 {
     EVP_PKEY *pkey;
-    int i;
-    if (ri->type == CMS_RECIPINFO_TRANS)
-        pkey = ri->d.ktri->pkey;
-    else if (ri->type == CMS_RECIPINFO_AGREE) {
-        EVP_PKEY_CTX *pctx = ri->d.kari->pctx;
+    EVP_PKEY_CTX *pctx = CMS_RecipientInfo_get0_pkey_ctx(ri);
+    EVP_PKEY_CTX *alloc_pctx = NULL;
+    int i, rv = 0;
+    X509_ALGOR **ppalg, *alg = NULL;
 
+    if (ri->type == CMS_RECIPINFO_TRANS) {
+        pkey = ri->d.ktri->pkey;
+        ppalg = &ri->d.ktri->keyEncryptionAlgorithm;
+    } else if (ri->type == CMS_RECIPINFO_AGREE) {
         if (pctx == NULL)
             return 0;
         pkey = EVP_PKEY_CTX_get0_pkey(pctx);
         if (pkey == NULL)
             return 0;
-    } else
+        ppalg = &ri->d.kari->keyEncryptionAlgorithm;
+    } else {
         return 0;
+    }
+
+    if (pkey->keymgmt == NULL)
+        goto legacy;
+
+    /* Compensate for not setting this up in the CMS_RECIPINFO_TRANS case */
+    if (pctx == NULL) {
+        if ((pctx = alloc_pctx = EVP_PKEY_CTX_new(pkey, NULL)) == NULL
+            || EVP_PKEY_encrypt_init(pctx) <= 0)
+            goto end;
```

Comment:
Should we raise some special error for this case?

Reply:
Because?  Mind you, there are error codes to say that this and that other library had an error and serves as a trace...  unfortunately, I recently discovered that those codes are messed up...